core/src/main/java/org/infinispan/protostream/impl/UnknownFieldSetImpl.java

```diff
@@ -198,15 +198,15 @@ public boolean hasTag(int tag) {
 
    @Override
    public void writeExternal(ObjectOutput out) throws IOException {
-      ByteArrayOutputStreamEx baos = new ByteArrayOutputStreamEx();
+      RandomAccessOutputStream baos = new RandomAccessOutputStreamImpl();
       TagWriter output = TagWriterImpl.newInstance(null, baos);
       writeTo(output);
       output.flush();
-      ByteBuffer buffer = baos.getByteBuffer();
-      int off = buffer.arrayOffset();
-      int len = buffer.limit() - off;
-      out.writeInt(len);
-      out.write(buffer.array(), off, len);
+
+      int pos = baos.getPosition();
+      out.writeInt(pos);
+      for (int i = 0; i < pos; i++)
+         out.write(baos.get(i));
```

Comment:
I took a closer look, and I suggest creating a method `void copyTo(DataOutput output) throws IOException;` in `RandomAccessOutputStream`

Writing byte by byte does not seem very efficient, mainly it has to check the capacity of the `DataOutput` on every invocation.